lib/pacemaker/pcmk_sched_promotable.c

```diff
@@ -272,131 +272,217 @@ sort_promotable_instance(gconstpointer a, gconstpointer b, gpointer data_set)
-        pcmk__colocation_t *constraint = (pcmk__colocation_t *) gIter->data;
+/*!
+ * \internal
+ * \brief Applies location preferences of resources with which the
+ *        promoted instance should be colocated.
+ *
+ * Merges the primary resource's node weights into this resource's
+ * node table.
+ *
+ * \param[in,out] rsc Clone wrapper.
+ *
+ * \note \p rsc is the dependent resource in these colocations.
+ */
+static void
+apply_promoted_dependent_colocations(pe_resource_t *rsc)
+{
+    for (GList *gIter = rsc->rsc_cons; gIter != NULL; gIter = gIter->next) {
+        pcmk__colocation_t *cons = (pcmk__colocation_t *) gIter->data;
+        uint32_t flags = (cons->score == INFINITY) ? pe_weights_none
+                                                   : pe_weights_rollback;
 
-        /* (re-)adds location preferences of resources that the
-         * master instance should/must be colocated with
-         */
-        if (constraint->role_lh == RSC_ROLE_MASTER) {
-            enum pe_weights flags = constraint->score == INFINITY ? 0 : pe_weights_rollback;
-
-            pe_rsc_trace(rsc, "RHS: %s with %s: %d", constraint->rsc_lh->id, constraint->rsc_rh->id,
-                         constraint->score);
-            rsc->allowed_nodes =
-                constraint->rsc_rh->cmds->merge_weights(constraint->rsc_rh, rsc->id,
-                                                        rsc->allowed_nodes,
-                                                        constraint->node_attribute,
-                                                        (float)constraint->score / INFINITY, flags);
+        if (cons->role_lh != RSC_ROLE_MASTER) {
+            continue;
         }
+
+        pe_rsc_trace(rsc, "RHS: %s with %s: %d", cons->rsc_lh->id,
+                     cons->rsc_rh->id, cons->score);
+
+        rsc->allowed_nodes =
+            cons->rsc_rh->cmds->merge_weights(cons->rsc_rh, rsc->id,
+                                              rsc->allowed_nodes,
+                                              cons->node_attribute,
+                                              (cons->score / (float) INFINITY),
+                                              flags);
     }
+}
 
-    gIter = rsc->rsc_cons_lhs;
-    for (; gIter != NULL; gIter = gIter->next) {
-        pcmk__colocation_t *constraint = (pcmk__colocation_t *) gIter->data;
+/*!
+ * \internal
+ * \brief Applies location preferences of resources that should be
+ *        colocated with the promoted instance.
+ *
+ * Merges the dependent resource's node weights into this resource's
+ * node table.
+ *
+ * \param[in,out] rsc Clone wrapper.
+ *
+ * \note \p rsc is the primary resource in these colocations.
+ */
+static void
+apply_promoted_primary_colocations(pe_resource_t *rsc)
+{
+    for (GList *gIter = rsc->rsc_cons_lhs; gIter != NULL; gIter = gIter->next) {
+        pcmk__colocation_t *cons = (pcmk__colocation_t *) gIter->data;
+        uint32_t flags = pe_weights_rollback | pe_weights_positive;
```

Comment:
might as well tag this const

Reply:
Meaning make the `pcmk__colocation_t *` a constant, or meaning something else? That would make sense.

Reply:
I meant flags, since the value is never changed. But cons could make sense too

Reply:
Done